File: Second Year/2nd Semester/Medii de Proiectare si Programare/Java/Laborator_1/src/main/java/repository/ParticipantDBRepository.java
package repository;

import domain.Participant;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ParticipantDBRepository implements ParticipantRepository {
    private JdbcUtils dbUtils;

    private static final Logger logger = LogManager.getLogger();

    public ParticipantDBRepository(Properties props){
        logger.info("Initializing ParticipantDBRepository with propertiesL {} ",props);
        dbUtils=new JdbcUtils(props);
    }

    @Override
    public Participant findOne(Integer integer) throws IllegalArgumentException {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        Participant participant = null;
        try(PreparedStatement preSmt= conn.prepareStatement("select * from Participants where id=?")){
            preSmt.setInt(1,integer);
            try(ResultSet resultSet = preSmt.executeQuery()){
                if(!resultSet.next())
                    return null;

                int id = resultSet.getInt("id");
                String first_name = resultSet.getString("first_name");
                String last_name = resultSet.getString("last_name");
                int points = resultSet.getInt("points");

                participant = new Participant(id,first_name,last_name,points);
            }
        }catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(participant);
        return participant;
    }

    @Override
    public Iterable<Participant> getAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Participant> participantList = new ArrayList<>();
        try(PreparedStatement preSmt=con.prepareStatement("select * from Participants")){
            try(ResultSet resultSet = preSmt.executeQuery()){
                while(resultSet.next()){
                    int id = resultSet.getInt("id");
                    String first_name = resultSet.getString("first_name");
                    String last_name = resultSet.getString("last_name");
                    int points = resultSet.getInt("points");
                    Participant participant = new Participant(id, first_name,last_name,points);
                    participantList.add(participant);
                }
            }
        } catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(participantList);
        return participantList;
    }

    @Override
    public void clear() {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        try(PreparedStatement preSmt= conn.prepareStatement("delete from Participants")){
            preSmt.executeUpdate();
        }catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit();
    }

    @Override
    public void save(Participant entity) {
        logger.traceEntry("saving task{} ", entity);
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into Participants (first_name,last_name,points) values (?,?,?)")){
            preStmt.setString(1,entity.getFirst_name());
            preStmt.setString(2,entity.getLast_name());
            preStmt.setInt(3,entity.getPoints());
            int result=preStmt.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB " + ex);
        }
        logger.traceExit();
    }

    @Override
    public void update(Participant entity) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        try(PreparedStatement preSmt= conn.prepareStatement("update Participants set first_name = ?, last_name = ?, points = ? where id=?")){
            preSmt.setString(1,entity.getFirst_name());
            preSmt.setString(2,entity.getLast_name());
            preSmt.setInt(3,entity.getPoints());
            preSmt.setInt(4,entity.getId());
            preSmt.executeUpdate();
        }catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(entity);
    }

    @Override
    public void delete(Integer id) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        try(PreparedStatement preSmt= conn.prepareStatement("delete from Participants where id=?")){
            preSmt.setInt(1,id);
            preSmt.executeUpdate();
        }catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(id);
    }

    @Override
    public List<Participant> findByActivityOrderedDescByPoints(String activity) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Participant> participantList = new ArrayList<>();
        try(PreparedStatement preSmt=con.prepareStatement("select * from Participants where activity = ? order by points desc")){
            preSmt.setString(1,activity);
            try(ResultSet resultSet = preSmt.executeQuery()){
                while(resultSet.next()){
                    int id = resultSet.getInt("id");
                    String first_name = resultSet.getString("first_name");
                    String last_name = resultSet.getString("last_name");
                    int points = resultSet.getInt("points");
                    Participant participant = new Participant(id, first_name,last_name,points);
                    participantList.add(participant);
                }
            }
        } catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(participantList);
        return participantList;
    }
}

File: Second Year/2nd Semester/Medii de Proiectare si Programare/C#/Laborator_3/repository/ParticipantDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using Laborator_1.domain;
using log4net;

namespace Laborator_1.repository
{
    public class ParticipantDBRepository:ParticipantRepository
    {
        private static readonly ILog log = LogManager.GetLogger("ParticipantDBRepository");
        private IDictionary<string, string> props;

        public ParticipantDBRepository(IDictionary<string, string> props)
        {
            log.Info("Creating ParticipantDBRepository");
            this.props = props;
        }
        
        public Participant FindOne(int id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection conn = DBUtils.getConnection(props);
            using (var command = conn.CreateCommand())
            {
                command.CommandText = "select id,first_name,last_name,points  from Participants where id=@id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                using (var dataR = command.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idV = dataR.GetInt32(0);
                        String first_name = dataR.GetString(1);
                        String last_name = dataR.GetString(2);
                        int points = dataR.GetInt32(3);
                        Participant participant = new Participant(idV, first_name, last_name, points);
                        log.InfoFormat("Existing findOne with value {0}", participant);
                        return participant;
                    }
                }
            }
            log.InfoFormat("Existing findOne with value {0}", null);
            return null;
        }

        public IEnumerable<Participant> FindAll()
        {
            IDbConnection conn = DBUtils.getConnection(props);
            IList<Participant> participants = new List<Participant>();
            using (var command = conn.CreateCommand())
            {
                command.CommandText = "select id,first_name,last_name,points  from Participants";
                using (var dataR = command.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int id = dataR.GetInt32(0);
                        String first_name = dataR.GetString(1);
                        String last_name = dataR.GetString(2);
                        int points = dataR.GetInt32(3);
                        Participant participant = new Participant(id, first_name, last_name, points);
                        participants.Add(participant);
                    }
                }
            }
            return participants;
        }

        public void Save(Participant entity)
        {
            var conn = DBUtils.getConnection(props);
            using (var command = conn.CreateCommand())
            {
                command.CommandText =
                    "insert into Participants(first_name,last_name,points) values(@first_name, @last_name, @points)";

                var paramFirstName = command.CreateParameter();
                paramFirstName.ParameterName = "@first_name";
                paramFirstName.Value = entity.First_Name;
                command.Parameters.Add(paramFirstName);
                
                var paramLastName = command.CreateParameter();
                paramLastName.ParameterName = "@last_name";
                paramLastName.Value = entity.Last_Name;
                command.Parameters.Add(paramLastName);
                
                var paramPoints = command.CreateParameter();
                paramPoints.ParameterName = "@points";
                paramPoints.Value = entity.Points;
                command.Parameters.Add(paramPoints);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No participant added!");
            }
        }

        public void Delete(int id)
        {
            IDbConnection conn = DBUtils.getConnection(props);
            using (var command = conn.CreateCommand())
            {
                command.CommandText = "delete from Participants where id=@id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);
                var dataR = command.ExecuteNonQuery();
                if (dataR == 0)
                    throw new RepositoryException("No participant deleted!");
            }
        }

        public void Update(Participant entity)
        {
            var conn = DBUtils.getConnection(props);
            using (var command = conn.CreateCommand())
            {
                command.CommandText =
                    "update Participants set first_name = @first_name, last_name = @last_name, points = @points where id = @id";
                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = entity.Id;
                command.Parameters.Add(paramId);

                var paramFirstName = command.CreateParameter();
                paramFirstName.ParameterName = "@first_name";
                paramFirstName.Value = entity.First_Name;
                command.Parameters.Add(paramFirstName);
                
                var paramLastName = command.CreateParameter();
                paramLastName.ParameterName = "@last_name";
                paramLastName.Value = entity.Last_Name;
                command.Parameters.Add(paramLastName);
                
                var paramPoints = command.CreateParameter();
                paramPoints.ParameterName = "@points";
                paramPoints.Value = entity.Points;
                command.Parameters.Add(paramPoints);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No participant updated!");
            }
        }

        public List<Participant> FindByActivityOrderedDescByPoints(string Activity)
        {
            log.InfoFormat("Entering findByActivityOrderedDescByPoints with value {0}", Activity);
            IDbConnection conn = DBUtils.getConnection(props);
            List<Participant> participants = new List<Participant>();
            using (var command = conn.CreateCommand())
            {
                command.CommandText = "select * from Participants where activity=@activity order by points desc";
                IDbDataParameter paramActivity = command.CreateParameter();
                paramActivity.ParameterName = "@activity";
                paramActivity.Value = Activity;
                command.Parameters.Add(paramActivity);

                using (var dataR = command.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int id = dataR.GetInt32(0);
                        String first_name = dataR.GetString(1);
                        String last_name = dataR.GetString(2);
                        int points = dataR.GetInt32(3);
                        Participant participant = new Participant(id, first_name, last_name, points);
                        log.InfoFormat("Existing findByActivityOrderedDescByPoints with balue {0}",Activity);
                        participants.Add(participant);
                    }
                }
            }
            log.InfoFormat("Existing findByActivityOrderedDescByPoints with value {0}",Activity);
            return participants;
        }
    }
}